### PR TITLE
[Enabler][1104]migrate_zos_backup_restore

### LINKED ISSUE #1104: [Enabler] [Dev] [zos_backup_restore] Migrate module, module utilities and test cases to ZOAU 1.3
## Ansible module
`zos_backup_restore`

### See Epic
[[Enabler] [Epic] Adopt new ZOAU 1.3 Python changes](https://github.com/ansible-collections/ibm_zos_core/issues/931)

### Enabler description

With the adoption of ZOAU 1.3, modules, module utilities and test cases may need to be updated according to the [migration guide](https://www.ibm.com/docs/en/zoau/1.3.x?topic=planning-migrating-zoau-v13-from-v12x-earlier). 

APIs changes affecting module:
- Impacted by ZOAU Dataset API: Yes
- Impacted by ZOAU jobs API: No
- Impacted by the rename of types.py to ztypes.py: No
- Impacted by the contents of module volumes.py having been moved to datasets.py: No
- Impacted by opercmd timeout to milli seconds: No
- Impacted by JSON schema change in either apfadm, dmod, dsed or zinfo: No
- Impacted by Exceptions schema: Yes but the code is not affected because the portion of the schema accessed by the module has not changed. 

Review the [zoautil_py datasets module migration doc](https://www.ibm.com/docs/en/zoau/1.3.x?topic=planning-migrating-zoau-v13-from-v12x-earlier#changes-to-the-datasets-module).

Upon review, this module is affected by:
- the `dataset` class being relocated to zoautil_py.datasets.Dataset.
   - See `from zoautil_py import datasets, exceptions`
- the function datasets.zip() is renamed to datasets.dzip().
   - see calls to `datasets.zip(**zoau_args)`
- function datasets.unzip() is renamed to datasets.dunzip().
   - see calls to `response = datasets._unzip(**zoau_args)`
- the function datasets.hlq() is renamed to datasets.get_hlq().
  - see calls to `hlq = datasets.hlq()`

This issue depends on shared import identified by git issues:
- [module_utils/better_arg_parser.py](https://github.com/ansible-collections/ibm_zos_core/issues/1116)
- [module_utils/data_set.py](https://github.com/ansible-collections/ibm_zos_core/issues/1118)

### PR DESCRIPTION
**Pull request by AndreMarcel99**
##### SUMMARY
Migrate module zos backup and restore to use ZOAU 1.3 and add the option of temporal hight level qualifier for the user.

Fixes #1104 and #1262 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
Migrate functionality of zos backup an restore to ZOAU 1.3, add option of tmphlq to the module and remove all the test cases could left adding a new option.

##### ADDITIONAL INFORMATION
Captura de pantalla 2024-03-01 a la(s) 10 33 09 a m

Captura de pantalla 2024-02-29 a la(s) 9 48 40 a m
372-61d58e264e4b)

### CODE REVIEW

**Review by richp405 — APPROVED** [review]

File: tests/functional/modules/test_zos_backup_restore.py [hunk 1]
```diff
@@ -73,6 +73,10 @@ def delete_data_set(hosts, data_set_name):
 def delete_file(hosts, path):
     hosts.all.file(path=path, state="absent")
 
+def delete_remnants(hosts):
+    hosts.all.shell(cmd="drm 'ANSIBLE.*'")
+    hosts.all.shell(cmd="drm 'TEST.*'")
```
Comment (fernandofloresg) [inline tests/functional/modules/test_zos_backup_restore.py:78]: Question about this TEST.*, which data sets get deleted by it ? 
Also, why is this needed if tests do delete the data sets in their finally section? Is the module leaving temporary data sets behind?

Comment (AndreMarcel99) [reply tests/functional/modules/test_zos_backup_restore.py:78]: For the test `test_backup_and_restore_of_data_set` is fixed by adding a new hlq to ensure works properly and do not mesh with the dataset is not removed as the previous behavior, and yes as my changes for random dataset name make it almost impossible to found all datasets to be remove with a specific name

Comment (fernandofloresg) [reply tests/functional/modules/test_zos_backup_restore.py:78]: makes sense

**Review by fernandofloresg — APPROVED** [review]

**Review by rexemin — APPROVED** [review]

File: plugins/modules/zos_backup_restore.py [hunk 2]
```diff
@@ -482,10 +497,11 @@ def backup(
         space_type (str): The unit of measurement to use when defining data set space.
         sms_storage_class (str): Specifies the storage class to use.
         sms_management_class (str): Specifies the management class to use.
+        tmp_hlq (str): Specifies the tmp hlq to temporary datasets
     """
     args = locals()
     zoau_args = to_dzip_args(**args)
-    datasets.zip(**zoau_args)
+    datasets.dzip(**zoau_args)
```
Comment (ketankelkar) [inline plugins/modules/zos_backup_restore.py:504]: I'm curious why the `dunzip` operation in `restore()` below is wrapped in a try/except but this one isn't.

Comment (AndreMarcel99) [reply plugins/modules/zos_backup_restore.py:504]: Is to keep behavior because zip and zip if fails throw error. But below dzip check if the rc within 1 and 4.

File: changelogs/fragments/1265_Migrate_zos_backup_restore.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+trivial:
+  - zos_backup_restore - Refactor zos_backup_restore module and functional tests for ZOAU v1.3.0.
+    (https://github.com/ansible-collections/ibm_zos_core/pull/1265).
+minor_changes:
+  - zos_backup_restore - Add tmp_hlq option to the user interface to override the default high level qualifier
+    (HLQ) for temporary and backup.
```
Comment (ketankelkar) [inline changelogs/fragments/1265_Migrate_zos_backup_restore.yml:6]: This sounds more complete to me:
`(HLQ) for temporary and backup.` --> `(HLQ) for temporary and backup data sets.`

File: plugins/modules/zos_backup_restore.py [hunk 4]
```diff
@@ -523,23 +540,26 @@ def restore(
...
-    true_rc = response.rc
-    if response.rc > 0:
-        output = response.stdout_response + response.stderr_response
-        true_rc = get_real_rc(output) or true_rc
-    if true_rc > 0 and true_rc <= 4:
+    if rc > 0 and rc <= 4:
         if recover is not True:
             failed = True
-    elif true_rc > 0:
+    elif rc > 0:
```
Comment (ketankelkar) [inline plugins/modules/zos_backup_restore.py:558]: not a show-stopper, shouldn't this one be `elif rc>4`? Since the rc 1-4 case is covered by the if?
